File: python/gpt/p136_nc.py
# Python (Flask)
# NOTE: In real systems, use POST (not URL/query params) + HTTPS to avoid credential leakage via logs, caches, referrers.

from flask import Flask, request, Response
import time
import hmac
import hashlib
from collections import defaultdict, deque

app = Flask(__name__)

# Demo credentials (store hashes, not plaintext)
EXPECTED_USERNAME = "admin"
SALT = b"static-demo-salt-change-me"
ITERATIONS = 200_000
DKLEN = 32

# "password123" -> derived key (demo)
_EXPECTED_PW_DK = hashlib.pbkdf2_hmac("sha256", b"password123", SALT, ITERATIONS, dklen=DKLEN)

# Simple in-memory rate limiter: 5 requests per 60 seconds per IP
WINDOW_SEC = 60
MAX_REQ = 5
_hits = defaultdict(deque)

def _rate_limited(ip: str) -> bool:
    now = time.time()
    q = _hits[ip]
    while q and now - q[0] > WINDOW_SEC:
        q.popleft()
    if len(q) >= MAX_REQ:
        return True
    q.append(now)
    return False

@app.get("/check_mod")
def check_mod():
    ip = request.headers.get("X-Forwarded-For", request.remote_addr) or "unknown"
    if _rate_limited(ip):
        return Response("false", status=429, mimetype="text/plain")

    username = request.args.get("username", "")
    password = request.args.get("password", "")

    # Constant-time comparisons (avoid leaking which field was wrong)
    user_ok = hmac.compare_digest(username, EXPECTED_USERNAME)

    pw_dk = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), SALT, ITERATIONS, dklen=DKLEN)
    pw_ok = hmac.compare_digest(pw_dk, _EXPECTED_PW_DK)

    ok = user_ok and pw_ok
    return Response("true" if ok else "false", status=200 if ok else 401, mimetype="text/plain")

if __name__ == "__main__":
    app.run(host="127.0.0.1", port=5000, debug=False)

File: javascript/gpt/p136_nc.js
// JavaScript (Node.js + Express)
// NOTE: In real systems, use POST (not URL/query params) + HTTPS to avoid credential leakage via logs, caches, referrers.

const express = require("express");
const crypto = require("crypto");

const app = express();

const EXPECTED_USERNAME = "admin";
const SALT = Buffer.from("static-demo-salt-change-me");
const ITERATIONS = 200000;
const DKLEN = 32;
const DIGEST = "sha256";

// Demo expected password = "password123"
const EXPECTED_PW_DK = crypto.pbkdf2Sync("password123", SALT, ITERATIONS, DKLEN, DIGEST);

// Simple in-memory rate limiter: 5 requests per 60 seconds per IP
const WINDOW_MS = 60_000;
const MAX_REQ = 5;
const hits = new Map(); // ip -> array of timestamps (ms)

function rateLimited(ip) {
  const now = Date.now();
  const arr = hits.get(ip) || [];
  while (arr.length && now - arr[0] > WINDOW_MS) arr.shift();
  if (arr.length >= MAX_REQ) {
    hits.set(ip, arr);
    return true;
  }
  arr.push(now);
  hits.set(ip, arr);
  return false;
}

function safeEqualString(a, b) {
  // timingSafeEqual requires same-length buffers; normalize to fixed length
  const A = Buffer.from(String(a), "utf8");
  const B = Buffer.from(String(b), "utf8");
  const len = Math.max(A.length, B.length, 1);
  const aP = Buffer.alloc(len);
  const bP = Buffer.alloc(len);
  A.copy(aP);
  B.copy(bP);
  return crypto.timingSafeEqual(aP, bP) && A.length === B.length;
}

app.get("/check_mod", (req, res) => {
  const ip = (req.headers["x-forwarded-for"] || req.socket.remoteAddress || "unknown").toString().split(",")[0].trim();

  if (rateLimited(ip)) {
    res.status(429).type("text").send("false");
    return;
  }

  const username = (req.query.username ?? "").toString();
  const password = (req.query.password ?? "").toString();

  // Constant-time comparisons (avoid leaking which field was wrong)
  const userOk = safeEqualString(username, EXPECTED_USERNAME);

  const pwDk = crypto.pbkdf2Sync(password, SALT, ITERATIONS, DKLEN, DIGEST);
  const pwOk = crypto.timingSafeEqual(pwDk, EXPECTED_PW_DK);

  const ok = userOk && pwOk;
  res.status(ok ? 200 : 401).type("text").send(ok ? "true" : "false");
});

app.listen(3000, "127.0.0.1");